tests/unit/test_braille.py

```diff
@@ -0,0 +1,96 @@
+		self.fakeDesktopObj=NVDAObjectWithRole(role=controlTypes.ROLE_WINDOW)
+		api.setDesktopObject(self.fakeDesktopObj)
+		# Forcefully create a fake focus ancestry
+		globalVars.focusAncestors=[api.getDesktopObject(),NVDAObjectWithRole(role=controlTypes.ROLE_DIALOG),NVDAObjectWithRole(role=controlTypes.ROLE_LIST)]
+		braille.handler.handleGainFocus(self.obj)
+		# Make sure that we are testing with three regions
+		self.assertEqual(len(self.regionsWithPositions),3)
+
+	def test_fillDisplay(self):
+		"""Test for the case where both the focus object and its ancestors should be visible on a 40 cell display."""
+		conf['braille']['focusContextPresentation']="fill"
+		# Since we set the presentation mode, simulate another gainFocus so the regions will be updated properly
+		braille.handler.handleGainFocus(self.obj)
+		# WindowEndPos should be retrieved before we attempt to get the start position
+		# This is because getting windowEndPos can update windowStartPos
+		# Both the focus object and its parent should fit on the display
+		# Thus, the window end position is equal to the end position of the 3rd region
+		self.assertEqual(braille.handler.buffer.windowEndPos,self.regionsWithPositions[2][2])
+		# The start position should be 0 now
+		self.assertEqual(braille.handler.buffer.windowStartPos,0)
+
+	def test_scrollOnly(self):
+		"""Test for the case where the focus object should be visible hard left on a display."""
+		conf['braille']['focusContextPresentation']="scroll"
+		braille.handler.handleGainFocus(self.obj)
+		# Only the focus object should be visible on the display
+		# This means that the window end position is equal to the end position of the 3rd region
+		self.assertEqual(braille.handler.buffer.windowEndPos,self.regionsWithPositions[2][2])
+		# This also means that the window start position is equal to the start position of the 3rd region
+		self.assertEqual(braille.handler.buffer.windowStartPos,self.regionsWithPositions[2][1])
+
+	def test_hybrid(self):
+		"""Test for the case where the focus object as well as ancestry differences should be visible on the display"""
+		conf['braille']['focusContextPresentation']="hybrid"
+		# Clean up the cached ancestry regions
+		braille.invalidateCachedFocusAncestors(0)
+		# Regenerate the regions
+		braille.handler.handleGainFocus(self.obj)
+		# Both the focus object and its parents should be visible, equivalent to always fill display
+		self.assertEqual(braille.handler.buffer.windowEndPos,self.regionsWithPositions[2][2])
+		self.assertEqual(braille.handler.buffer.windowStartPos,0)
+		# Do another focus to simulate a new focus object with equal ancestry
+		braille.handler.handleGainFocus(self.obj)
+		# Only the focus object should be visible now, equivalent to scroll only
+		self.assertEqual(braille.handler.buffer.windowEndPos,self.regionsWithPositions[2][2])
+		self.assertEqual(braille.handler.buffer.windowStartPos,self.regionsWithPositions[2][1])
+		# Clean up the cached focus ancestors

```

Comment:
You're only cleaning up ancestors to a certain point, so I think this comment needs some clarification. You could use the roles to help explain.